src/GitHub.InlineReviews/Services/PullRequestSessionService.cs

```diff
@@ -0,0 +1,142 @@
+            return repo.Head.Tip.Sha;
+        }
+
+        /// <inheritdoc/>
+        public async Task<bool> IsUnmodifiedAndPushed(ILocalRepositoryModel repository, string relativePath, byte[] contents)
+        {
+            var repo = GetCachedRepository(repository);
+            return !await gitClient.IsModified(repo, relativePath, contents) &&
+                   await gitClient.IsHeadPushed(repo);
+        }
+
+        public async Task<byte[]> ExtractFileFromGit(
+            ILocalRepositoryModel repository,
+            int pullRequestNumber,
+            string sha,
+            string relativePath)
+        {
+            var repo = GetCachedRepository(repository);
+
+            try
+            {
+                return await gitClient.ExtractFileBinary(repo, sha, relativePath);
+            }
+            catch (FileNotFoundException)
+            {
+                var pullHeadRef = $"refs/pull/{pullRequestNumber}/head";
+                await gitClient.Fetch(repo, "origin", sha, pullHeadRef);
+                return await gitClient.ExtractFileBinary(repo, sha, relativePath);
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task<byte[]> ReadFileAsync(string path)
+        {
+            if (File.Exists(path))
+            {
+                try
+                {
+                    using (var file = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, true))
+                    {
+                        var buffer = new MemoryStream();
+                        await file.CopyToAsync(buffer);
+                        return buffer.ToArray();
+                    }
+                }
+                catch { }
+            }
+
+            return null;
+        }
+
+        /// <inheritdoc/>
+        public async Task<string> GetPullRequestMergeBase(ILocalRepositoryModel repository, IPullRequestModel pullRequest)
+        {
+            var baseSha = pullRequest.Base.Sha;
+            var headSha = pullRequest.Head.Sha;
+            var key = new Tuple<string, string>(baseSha, headSha);
+
+            string mergeBase;
+            if(mergeBaseCache.TryGetValue(key, out mergeBase))
+            {
+                return mergeBase;
+            }
+
+            var repo = GetCachedRepository(repository);
+            var remote = await gitClient.GetHttpRemote(repo, "origin");
+
+            var baseRef = pullRequest.Base.Ref;
+            var pullNumber = pullRequest.Number;
+            mergeBase = await gitClient.GetPullRequestMergeBase(repo, remote.Name, baseSha, headSha, baseRef, pullNumber);
+            if (mergeBase == null)
+            {
+                throw new FileNotFoundException($"Couldn't find merge base between {baseSha} and {headSha}.");
+            }
+
+            mergeBaseCache[key] = mergeBase;
+            return mergeBase;
+        }
+
+        IRepository GetCachedRepository(ILocalRepositoryModel model)
```

Comment:
Are we sure this will be safe? The `GitClient` uses a threadpool thread for executing its operation, meaning that an essientially random thread will be used. Given that `Repository` cannot be shared between threads, this looks likely to me to cause problems. Am I missing anything?